fixes #1459 - ipxe provisioning support

timogoebel opened the pull request:
This commit adds better iPXE support and the concept of the "global default template" and the "local boot template".

A host can request `http://foreman.example.com/unattended/iPXE`. If the host is unknown to Foreman, the `global default template` is rendered. If the host is known and is in build mode the iPXE provisioning template is rendered. If the host is not in build mode a `iPXE local boot` template is rendered. This way a user can leverage the workflow known from pxelinux with iPXE. This is especially helpful if the dhcp and tftp servers are not managed by Foreman.

This is a port of https://github.com/ananace/foreman_ipxe to core.

cc: @lzap, @ananace

timogoebel committed:
fixes #1459 - ipxe provisioning support

ananace commented:
I'm sitting here thinking if the provided global template should maybe be rewritten to better match the other global default templates; split into a pair of chainload and discovery snippets, along with generation of profile boot options.

For the chainload, you can choose to either just return from the boot agent - `exit` - or explicitly boot a drive - `sanboot --no-describe --drive 0x80` (0x81 and 0x82 are also possible common targets).

On the other hand, since iPXE is written as an imperative script instead of just configuration, such snippets would probably be a bit messier to deal with, probably resulting in something like;

```erb
<%#
kind: iPXE
name: iPXE global default
model: ProvisioningTemplate
-%>
#!ipxe

set menu-default discovery
set menu-timeout 5000

:menu
menu "Global default"

<%= snippet 'ipxe_chainload' %>

<%= snippet 'ipxe_discovery' %>

<% unless @profiles.nil? -%>
item --gap -- ---- Profiles ----
<% for profile in @profiles
  url = default_template_url(profile[:template], profile[:hostgroup])
  case profile[:pxe_type]
  when 'kickstart'
    append = "ks=#{url} ksdevice=bootif network kssendmac ks.sendmac inst.ks.sendmac"
  when 'preseed'
    locale = profile[:hostgroup].params['lang'] || 'en_US'
    append = "interface=auto url=#{url} ramdisk_size=10800 root=/dev/rd/0 rw auto hostname=unassigned-hostname locale=#{locale} console-setup/ask_detect=false console-setup/layout=USA console-setup/variant=USA"
  else
    raise("PXE type #{profile[:pxe_type]} not supported by template #{template_name}")
  end %>
item boot_profile_<%= #... %> <%= "#{profile[:hostgroup]} - #{profile[:template]}" %>
goto :post_profile_<%= #... %>

:boot_profile <%= #... %>
kernel <%= profile[:kernel] %> <%= append %>
initrd <%= profile[:initrd] %>
boot || goto failed
goto menu

:post_profile_<%= #... %>
<% end %>
<% end -%>

choose --timeout ${menu-timeout} --default ${menu-default} selected || goto cancel
set menu-timeout 0
goto ${selected}

:cancel
echo Menu canceled, dropping to shell

:shell
echo Use the command 'exit' to return to menu
shell
set menu-timeout 0
goto start

:failed
echo Boot failed, dropping to shell
goto shell
```

```erb
<%#
kind: snippet
name: ipxe_chainload
model: ProvisioningTemplate
snippet: true
-%>
item --gap --      ---- Chainload ----
item exit          Exit iPXE and continue boot
item sanboot_80    Boot drive 0x80
item sanboot_81    Boot drive 0x81
item sanboot_82    Boot drive 0x82
goto menu_post_chainload

:exit
exit

:sanboot_80
sanboot --no-describe --drive 0x80 || goto failed
goto menu

:sanboot_81
sanboot --no-describe --drive 0x81 || goto failed
goto menu

:sanboot_82
sanboot --no-describe --drive 0x82 || goto failed
goto menu

:menu_post_chainload
```

```erb
<%#
kind: snippet
name: ipxe_discovery
model: ProvisioningTemplate
snippet: true
-%>
item --gap --          ---- Discovery ----
item --key d discovery Boot discovery image
goto menu_post_discovery

:discovery
dhcp
kernel ${next-server}/boot/fdi-image/vmlinuz0 rootflags=loop root=live:/fdi.iso rootfstype=auto ro rd.live.image acpi=force rd.luks=0 rd.md=0 rd.dm=0 rd.lvm=0 rd.bootif=0 rd.neednet=0 nomodeset proxy.url=<%= foreman_server_url %> proxy.type=foreman BOOTIF=01-${net0/mac}
initrd ${next-server}/boot/fdi-image/initrd0.img
boot || goto failed
goto menu

:menu_post_discovery
```

lzap commented on app/controllers/unattended_controller.rb:
```diff
@@ -11,14 +11,16 @@ class UnattendedController < ApplicationController
   before_action :get_host_details, :except => [:hostgroup_template, :built, :failed]
   before_action :get_built_host_details, :only => [:built, :failed]
-  before_action :allowed_to_install?, :except => :hostgroup_template
+  before_action :allowed_to_install?, :except => :hostgroup_template, :unless => -> { ipxe_request? }
```
This is hacky, can you move the logic into `allowed_to_install?` method itself?

Also, this opens doors to all iPXE requests while we only want to allow the initial request. Can we identify the initial (bootstrap) request? How about passing an extra option from DHCP filename option that would do this? Something like `/unattended/iPXE?bootstrap=yes` or we can reuse `token=boostrap` perhaps.

lzap commented on app/services/foreman/ipxe/message_renderer.rb:
```diff
@@ -0,0 +1,21 @@
+      end
+    end
+  end
+end
```
This is nice indeed, message on screen is great. But seriously, 21 lines and a class to generate three lines? Why not to do it in the `unattended_controller.rb`? Ignore the metric if there's complain, I do not take them seriously.

lzap replied:
Oh after more lines I see the `StubRenderer` this is fine then. Although this could be all inline... :)

lzap commented on app/controllers/unattended_controller.rb:
```diff
@@ -11,14 +11,16 @@ class UnattendedController < ApplicationController
   before_action :get_host_details, :except => [:hostgroup_template, :built, :failed]
   before_action :get_built_host_details, :only => [:built, :failed]
-  before_action :allowed_to_install?, :except => :hostgroup_template
+  before_action :allowed_to_install?, :except => :hostgroup_template, :unless => -> { ipxe_request? }
```
This is hacky, can you move the logic into `allowed_to_install?` method itself?

Also, this opens doors to all iPXE requests while we only want to allow the initial request. Can we identify the initial (bootstrap) request? How about passing an extra option from DHCP filename option that would do this? Something like `/unattended/iPXE?bootstrap=yes` or we can reuse `token=boostrap` perhaps.

timogoebel replied:
Yeah, this is indeed hacky. I just ported the plugin code, still need to clean this up.

I don't get what you want to achieve with this bootstrap parameter. What's the aim?
I want to support dumb DHCP servers with this that do not support any conditional logic or different options for specific hosts.

lzap replied:
Let me explain, and maybe I am missing your iPXE use case. But in deployments I've worked with, because DHCP is dumb you have to boot what I call "bootstrap ipxe" script. That's a URL that never changes and contains static ipxe script. In that script, you chainboot another ipxe script but now you can use scripting capabilities therefore you can say `/unattended/iPXE?mac=${net0/mac}`.

I am under impression that the global template is actually what I call the bootstrap. I was thinking creating some extra parameter saying "get me the global template" would allow us doing cleaner code. But it can also be as simple as without any `mac` or `token` or anaconda MAC HTTP header  just return the global template. That's also an option.

Sorry I often think loud and this confuses people, I need to stop.

timogoebel replied:
It's alright. You actually make a valid point.
In my use case I currently have an embedded iPXE template that loads the global template from Foreman.

```
#!ipxe
dhcp
chain http://tfmdemo.beispiel.xyz/unattended/iPXE?mac=${net0/mac}
```

You are right, we need this as well. But I think we should add this as a separate controller action (in a separate commit).

lzap commented on app/controllers/unattended_controller.rb:
```diff
@@ -69,8 +72,51 @@ def preview?
+      unless template
+        render_ipxe_message(message: _("Global iPXE template '%s' not found") % name)
+        return
+      end
```
I'd use a "bootstrap" parameter for that, it's good to be explicit if possible.

lzap commented on app/controllers/unattended_controller.rb:
```diff
@@ -11,14 +11,16 @@ class UnattendedController < ApplicationController
   # maximum size of built/failed request body accepted to prevent DoS (in bytes)
   MAX_BUILT_BODY = 65535
 
+  IPXE_TEMPLATE_PARAMETER = 'ipxe_template'
```
Why this is a constant? I'd go with inline string and let the compiler to optimize this.

lzap commented on app/controllers/unattended_controller.rb:
```diff
@@ -69,8 +72,51 @@ def preview?
+        template: Foreman::Ipxe::StubRenderer.new(
+          message: 'iPXE default local boot fallback'
+        ).to_s
+      )
```
Ditto for this block.

lzap commented on app/controllers/unattended_controller.rb:
```diff
@@ -69,8 +72,51 @@ def preview?
+        render_ipxe_message(message: _("iPXE template '%s' specified via a host parameter not found") % name)
+        return
+      end
+    end
```
So parameter `ipxe_template` must be always present. Why not to have some predefined template so this works out of box? Edit: Oh I see it down below, this is just an "extra" template overriding. Do we need this? Honestly nobody will ever use that until we advertise this. I'd rather have this working globally, but even then users will struggle finding this feature, myself included. But I am fine with having this if you rely on that.

timogoebel replied:
I'd be fine with removing this for the initial implementation.

lzap commented on app/controllers/unattended_controller.rb:
```diff
@@ -51,6 +53,7 @@ def hostgroup_template
   # i.e. /unattended/provision will render the provisioning template for the requesting host
   def host_template
     return head(:not_found) unless params[:kind].present?
+    return render_ipxe_template if ipxe_request?
```
If there was an explicit flag support, an extra method could be only called for bootstrap options or when host parameter is present, otherwise normal workflow could follow. Not sure what ends up being cleaner code tho, you can leave this as is if it's good enough.

lzap commented on app/controllers/unattended_controller.rb:
```diff
@@ -69,8 +72,51 @@ def preview?
+    return unless verify_found_host
+    return unless allowed_to_install?
+
+    render_template('iPXE')
```
I think these four lines belong to a different method.

Both `verify_valid_host_token` and `verify_found_host` methods were actually already called from `get_host_details` before action so you don't need to call them IMHO. The `allowed_to_install?` should be extracted outside of scope of this method and another call of `render_template` might not be even necessary as this is already present in `host_template` controller action.

lzap commented on app/controllers/unattended_controller.rb:
```diff
@@ -92,6 +138,7 @@ def render_template(type)
+    return if ipxe_request?
     verify_valid_host_token
     verify_found_host
   end
```
I have to be missing something, you skip these checks and perform it in the newly created method. It reads awkward, maybe there is something I am missing here.

ananace replied:
Those verifications need to be skipped in case the request comes from a non-existant host.

When I wrote this code, I didn't want to have to do my own host discovery, but as the template rendering code also needs to trigger even if a host is *not* found - to support global template rendering - I had to skip the verifications at this point.

But yes, it reads a bit awkwardly.

lzap replied:
We should be running those validations in a single place rather than branching like that. You can easily tell when not to perform those (it is an iPXE request without mac, token or http mac parameter).

timogoebel replied:
@leewaa and I started refactoring the whole controller. Stay tuned. It's a lot of work but you it's clearly worth it.

lzap commented on app/models/concerns/pxe_loader_support.rb:
```diff
@@ -28,7 +28,10 @@ def all_loaders_map(precision = 'x64', httpboot_host = "httpboot_host")
+        "iPXE UEFI HTTP" => "http://#{httpboot_host}/httpboot/ipxe-#{precision}.efi",
+        "iPXE Chain BIOS" => "undionly.kpxe",
+        "iPXE Chain UEFI" => "ipxe.efi"
       }.freeze
```
Please also modify the `PXE_KINDS` hash above to match `undionly.kpxe` and `ipxe.efi`. Add few unit tests, we need to be sure mapping is correct from both directions.

lzap commented on app/controllers/unattended_controller.rb:
```diff
@@ -69,8 +72,51 @@ def preview?
+        name: 'iPXE default local boot fallback',
+        template: Foreman::Ipxe::StubRenderer.new(
+          message: 'iPXE default local boot fallback'
+        ).to_s
```
Can't we error out simply? What is the idea behind the stub template?

ananace replied:
This particular code was written before I'd set up a proper local boot template, so it could probably be removed now in favour of the real template instead.

lzap commented on app/views/unattended/provisioning_templates/iPXE/ipxe_default_local_boot.erb:
```diff
@@ -0,0 +1,9 @@
+-%>
+#!ipxe
+
+exit
```
You want to drop a comment explaining what this template does instead of the empty line? The purpose is I believe to skip booting from network and continue booting another device.

Chainbooting from MBR is I assume not possible in iPXE. And that would not work in EFI. Just thinking, ignore.

lzap commented on app/views/unattended/provisioning_templates/iPXE/ipxe_global_default.erb:
```diff
@@ -0,0 +1,47 @@
+#!ipxe
+
+set menu-default local
+set menu-timeout 5000
```
We have a setting for that. I think you might need to multiply it by 100 it's in seconds. Default to ten please.

    set menu-timeout <%= host_param('loader_timeout') * 100 || 1000 %>

ananace replied:
I think you mean multiply by 1000 :slightly_smiling_face:

lzap replied:
Yeah every single boot loader has a different time resolution, I've seen seconds, 1/10th, 1/100th... grrr.

timogoebel replied:
`@host` is not available in the global scope, so we cannot use this

lzap replied:
I'd swear I made this a global option, probably some local branch on my PC somewhere... Right, can't be done.

What you can do however is to use this from our global settings: `set menu-default <%= global_setting("default_pxe_item_global", "local") %>` just make sure menu items are named the same way as in PXEGrub2 template.

lzap commented on app/views/unattended/provisioning_templates/iPXE/ipxe_global_default.erb:
```diff
@@ -0,0 +1,47 @@
+:shell
+echo Use the command 'exit' to return to menu
+shell
+set menu-timeout 0
```
Why zero? This will disable the countdown completely? Not following, also there's the same line above.

ananace replied:
Most other PXE systems seem to act in similar manners, where the default boot timeout is aborted once you start interacting with the menu. Dropping into grub shell and then back to menu won't resume the default boot countdown as far as I recall.

lzap replied:
Okay.

lzap commented:
Nice, thanks. I like the idea of chanbooting via sanboot, if that fails iPXE can exit by default which will lead to fallback via BIOS when set.

ananace commented:
The only thing that needs to be considered is that iPXE is scripted instead of configured, which means the `menu` command must be executed before it's possible to start using the `item` command to set up menu options.
My really quick - basically pseudocode - example therefore uses code flow labels to allow for creating callable blocks without aborting the menu generation, which will require the main template to make sure that a menu is started - and ended - around the rendering of the snippets.

timogoebel commented:
* removed the fallback template, host parameter override
* refactored the unattended controller
* properly implemented ipxe loader support

leewaa commented on app/controllers/unattended_controller.rb:
```diff
@@ -50,7 +55,6 @@ def hostgroup_template
   # Generate an action for each template kind
   # i.e. /unattended/provision will render the provisioning template for the requesting host
   def host_template
-    return head(:not_found) unless params[:kind].present?
```
Its better to have this here rather than in a before action. Explicitness is king :)

timogoebel replied:
I agree, but this needs to happen before the method is called (that's why I had to move it to a before action).

leewaa reviewed (changes requested):
There are too many before actions now making it very hard to follow on what the code actually does. 

I think a separation of concerns is very important here by using one or more service objects. This way we can keep the logic in small, understandable, hunks of code which serve only one purpose each.

As requested by @timogoebel I will take a look at this and try to clean it up :)

kamils-iRonin commented on app/controllers/unattended_controller.rb:
```diff
@@ -63,14 +67,47 @@ def require_ssl?
+    return render_ipxe_message(message: message, status: status) if ipxe_request?
     # add a comment character (works with Red Hat and Debian systems) to avoid parsing errors
-    render(:plain => '# ' + _(error_message) % params, :status => status, :content_type => 'text/plain')
+    render(:plain => '# ' + message, :status => status, :content_type => 'text/plain')
```
you can use string interpolation for the message `"# #{message}"`

kamils-iRonin commented on app/controllers/unattended_controller.rb:
```diff
@@ -63,14 +67,47 @@ def require_ssl?
+    name = ProvisioningTemplate.global_template_name_for('iPXE')
+    template = ProvisioningTemplate.find_global_default_template(name, 'iPXE')
+
+    unless template
```
why not just
```
if template
  safe_render(template)
else
  render_ipxe_message(message: _("Global iPXE template '%s' not found") % name)
end
```

kamils-iRonin commented on app/controllers/unattended_controller.rb:
```diff
@@ -63,14 +67,47 @@ def require_ssl?
+    name = @host.local_boot_template_name(:ipxe) || ProvisioningTemplate.local_boot_name(:iPXE)
+    template = ProvisioningTemplate.find_by(name: name, template_kind: ipxe_template_kind)
+
+    unless template
```
the same as above

kamils-iRonin reviewed (commented):
looks good, I added minor comments

basiszwo commented on test/controllers/unattended_controller_test.rb:
```diff
@@ -13,19 +13,19 @@ class UnattendedControllerTest < ActionController::TestCase
     setup do
       ptable = FactoryBot.create(:ptable, :name => 'default',
-                                  :operatingsystem_ids => [operatingsystems(:redhat).id])
+                                 :operatingsystem_ids => [operatingsystems(:redhat).id])
```
why don't / can't you switch to current hash syntax if you are touching this anyway? (question applies to all other places as well)

timogoebel replied:
I usually do this, the old syntax gives me the creeps. Actually, I want to revert these changes to make the diff easier to read as this is clearly unrelated.

ekohl replied:
IIRC there's a rubocop for this. It's probably good to standardize on this in a separate issue.

leewaa replied:
Problem I see with making a change like this everywhere in a separate issue is it will cause allot of merge conflicts for existing issues. Hence I think the only plausible way without havoc is to change to the new syntax where/when ever changes are made.

lzap replied:
Agreed @leewaa but we tend to slowly enable various cops one by one in different patches. Our codebase is huge and there are many to go: https://github.com/theforeman/foreman/blob/develop/.rubocop_todo.yml

Actually not that many anymore, great job everyone!

Anyway, if cop does not resist, you are free to pick what you want. If you change to new syntax, not a problem. But this one particularly is hard to do in one commit I have to say.

ekohl replied:
@leewaa agreed one big PR is disrupting. I mostly meant by different issue that it shouldn't be part of this PR :)

leewaa replied:
Yes very difficult to do this with such a large code base.

lzap replied:
Anyway, can you revert those if you are not touching the lines?

timogoebel commented:
Ok, we have opened #6009 that does a lot of refactorings of the unattended controller. I'll go ahead and fix up the comments from @kamils-iRonin and I think then this is ready for a review.

timogoebel commented:
@lzap: Do you mind reviewing again?

The code is not perfect, but that's because unattended controller needed tidying up even before this PR. We have opened #6009 to address this, so please don't be overly strict here.

In addition, we'll open a follow up issue to add the possibility to provide an "iPXE bootstrap template" from Foreman. It could look something like this:

```
#!ipxe
dhcp
chain http://tfmdemo.beispiel.xyz/unattended/iPXE?mac=${net0/mac}
```

lzap commented on app/controllers/unattended_controller.rb:
```diff
@@ -8,11 +8,16 @@ class UnattendedController < ApplicationController
+  before_action :render_default_global_template, except: [:hostgroup_template, :built, :failed], if: -> { ipxe_request? && !@host }
+  before_action :render_local_boot_template, except: [:hostgroup_template, :built, :failed], if: -> { ipxe_request? && !@host.build? }
+  before_action :verify_valid_host_token, :except => [:hostgroup_template, :built, :failed]
+  before_action :verify_found_host, :except => :hostgroup_template
```
Regression - it used to be called via `:get_host_details, :except => [:hostgroup_template, :built, :failed]` and now it's missing the last two from the list. Intentional?

timogoebel replied:
Ouch, good catch. I seriously need to change all the except things to only.

lzap commented on test/controllers/unattended_controller_test.rb:
```diff
@@ -551,4 +551,114 @@ class UnattendedControllerTest < ActionController::TestCase
+        organization: tax_organization,
+        location: tax_location
+      )
+    end
```
Kudos for using let!

lzap commented on test/controllers/unattended_controller_test.rb:
```diff
@@ -551,4 +551,114 @@ class UnattendedControllerTest < ActionController::TestCase
+      test 'should render iPXE error when global iPXE template is not found' do
+        get :host_template, params: { kind: 'iPXE' }, session: set_session_user
+        assert_response :not_found
+        assert_includes @response.body, "Global iPXE template 'iPXE global default' not found"
```
Nitpick - do we want this error message to be presented in ipxe syntax like "echo ..."?

lzap commented on test/models/concerns/pxe_loader_support_test.rb:
```diff
@@ -151,6 +162,11 @@ def setup
       assert_equal "Grub UEFI", @subject.preferred_loader
     end
+
+    test "is PXEGrub for associated template iPXE" do
```
Wrong test description.

lzap reviewed (changes requested):
Looks cleaner, few nitpicks, but I consider this version good to go. There are relevant test failures tho.

If I was pushing too much I apologize, everytime I see unattended controller changes I do careful review with many ideas but the reason for that is - I worry about regressions. We don't want to leak root passwords and stuff. That's all.

timogoebel commented:
@lzap: Thanks for your very valuable comments. Fixed them and introduced `Operatingsystem.template_kinds_for_tftp` to fix the test failures.

lzap commented on app/controllers/unattended_controller.rb:
```diff
@@ -7,12 +7,17 @@ class UnattendedController < ApplicationController
-  before_action :handle_realm, :if => Proc.new { params[:kind] == 'provision' }
+  before_action :get_host_details, only: :host_template
+  before_action :get_built_host_details, only: [:built, :failed]
+  before_action :render_default_global_template, only: :host_template, if: -> { ipxe_request? && !@host }
```
Just curios - should we name this `render_default_global_ipxe_template` since it's ipxe only? The same for the line below.

leewaa replied:
This before action is removed/extracted in the refactorings of #6009 so I guess its ok like this ?

lzap replied:
Sure ok.

lzap commented on app/controllers/unattended_controller.rb:
```diff
@@ -92,27 +127,27 @@ def render_template(type)
   def get_built_host_details
     @host = find_host_by_spoof || find_built_host_by_token
     @host ||= find_host_by_ip_or_mac unless token_from_params.present?
-    verify_found_host
```
I am missing this call via before actions now, is that intentional?

leewaa replied:
The methods `get_host_details` and `get_built_host_details` are now both before actions...  `verify_found_host` and  `verify_valid_host_token` are also before actions... So no need to call them here anymore

lzap replied:
I get that, but this is how it used to be:

    before_action :get_host_details, :except => [:hostgroup_template, :built, :failed]
    before_action :get_built_host_details, :only => [:built, :failed]

Now compare except/only statements with this:

    before_action :verify_valid_host_token, only: :host_template
    before_action :verify_found_host, only: :host_template

For example get_built_host_details called verify_found_host for failed action but now it's called only for host_template action.

I am trying to find if we are not opening doors for attacks. Maybe I am just overlooking something, that's fine just show me.

timogoebel replied:
You are right, I messed up here. It should be

```ruby
before_action :verify_found_host, only: [:host_template, :built, :failed]
```

leewaa replied:
Ok 👍 clear reason why we want to remove these before actions 😄

lzap replied:
Yeah, happened to the best of us don't worry. This controller got out of control years ago... ;-)

lzap commented on app/models/concerns/pxe_loader_support.rb:
```diff
@@ -28,7 +29,10 @@ def all_loaders_map(precision = 'x64', httpboot_host = "httpboot_host")
         "Grub2 UEFI HTTPS" => "https://#{httpboot_host}/httpboot/grub2/grub#{precision}.efi",
         "Grub2 UEFI HTTPS SecureBoot" => "https://#{httpboot_host}/httpboot/grub2/shim#{precision}.efi",
-        "iPXE UEFI HTTP" => "http://#{httpboot_host}/httpboot/ipxe-#{precision}.efi"
+        "iPXE" => nil,
```
We need to find better name for this, this basically makes filename option to be rendered as `host.foreman_url('iPXE')`. So it is assumed that iPXE is somehow bootstrapped/embedded in the network card. Therefore I suggest "iPXE Embedded"? Can you also drop a comment:

    "iPXE Embedded" => nil, # renders directly as foreman_url('iPXE')

lzap commented on test/models/operatingsystem_test.rb:
```diff
@@ -340,4 +340,11 @@ class OperatingsystemTest < ActiveSupport::TestCase
+      host = FactoryBot.build(:host, :managed, pxe_loader: 'iPXE')
+      assert_equal 'http://foreman.some.host.fqdn/unattended/iPXE', host.operatingsystem.boot_filename(host)
+    end
+  end
```
Can you also add a test with Templates feature on host subnet on? That should render via Smart Proxy. Thanks.

lzap reviewed (changes requested):
One finding and few more nitpicks. Looks great.

timogoebel commented:
@lzap: I have addressed your comments, tests are green (if we pretend the intermittent failure does not exist). Do you mind taking another look? Thanks.

lzap commented on app/models/setting/provisioning.rb:
```diff
@@ -60,7 +60,7 @@ def self.default_settings
       self.set('clean_up_failed_deployment', N_("Foreman will delete virtual machine if provisioning script ends with non zero exit code"), true, N_('Clean up failed deployment')),
       self.set('name_generator_type', N_("Random gives unique names, MAC-based are longer but stable (and only works with bare-metal)"), 'Random-based', N_("Type of name generator"), nil, {:collection => Proc.new {NameGenerator::GENERATOR_TYPES} }),
-      self.set('default_pxe_item_global', N_("Default PXE (PXELinux, PXEGrub, PXEGrub2) menu item in global template - 'local', 'discovery' or custom"), 'local', N_("Default PXE global template entry")),
+      self.set('default_pxe_item_global', N_("Default PXE (PXELinux, PXEGrub, PXEGrub2, iPXE) menu item in global template - 'local', 'discovery' or custom"), 'local', N_("Default PXE global template entry")),
```
This conflicts with other PR I filed this morning, I am getting rid of the list of PXE loaders in this description text anyway. Can you simply remove this hunk? Sorry about that.

lzap reviewed (changes requested):
All good, thanks. There is one smallish nitpick.

Can you also rebase? Tests seem to fail for some weird taxonomy reason, develop is green today not sure: http://ci.theforeman.org/job/test_develop/

timogoebel commented:
@lzap: Rebased and removed the settings description.

lzap commented:
Figured out seed failures, the other one is known issue not yet fixed :-( 

https://github.com/theforeman/foreman/pull/6058

Started thread about the issue: https://community.theforeman.org/t/random-test-failures-fact-values-api-controller/11048

lzap commented:
I am good, can we get more approvals @kamils-iRonin @ekohl @leewaa ?

leewaa commented:
Yep, Good to merge!

timogoebel commented:
@lzap: I think you can safely hit the button, the refactorings PR is already open anyway :-)

mmoll commented:
merged, danke @timogoebel!

lzap commented:
Sorry folks, had to take two sick days but I am good. Looks like all is looking good here :-)
